facades/PC/build.gradle.kts

```diff
@@ -30,396 +28,370 @@ dateTimeFormat.timeZone = TimeZone.getTimeZone("UTC")
+    mainClass.set(extra.get("mainClassName") as String)
 }
 
 // Adjust as the Gradle 6 upgrade changed this path a bit
 sourceSets {
-    main.java.outputDir = new File("$buildDir/classes")
-    test.java.outputDir = new File("$buildDir/testClasses")
+    main { java.outputDir = File("$buildDir/classes") }
+    test { java.outputDir = File("$buildDir/testClasses") }
 }
 
 // Base the engine tests on the same version number as the engine
-version = project(':engine').version
-println "PC VERSION: $version"
+version = project(":engine").version
+logger.info("PC VERSION: {}", version)
 
 // Jenkins-Artifactory integration catches on to this as part of the Maven-type descriptor
-group = 'org.terasology.facades'
+group = "org.terasology.facades"
 
 dependencies {
-    implementation project(':engine')
-    implementation group: 'org.reflections', name: 'reflections', version: '0.9.10'
+    implementation(project(":engine"))
+    implementation(group = "org.reflections", name = "reflections", version = "0.9.10")
 
     // TODO: Consider whether we can move the CR dependency back here from the engine, where it is referenced from the main menu
-    implementation group: 'org.terasology.crashreporter', name: 'cr-terasology', version: '4.1.0'
+    implementation(group = "org.terasology.crashreporter", name = "cr-terasology", version = "4.1.0")
 }
 
 // Instructions for packaging a jar file for the PC facade
-jar {
+tasks.named<Jar>("jar") {
     manifest {
         //TODO: Maybe later add the engine's version number into here?
-        def manifestClasspath = configurations.runtimeClasspath.collect { it.getName() }.join(" ")
-        attributes("Main-Class": mainClassName)
-        attributes("Class-Path" : manifestClasspath)
-        attributes("Implementation-Title": "Terasology-" + project.name)
-        attributes("Implementation-Version": env.BUILD_NUMBER + ", " + env.GIT_BRANCH + ", " + env.BUILD_ID)
+        attributes["Main-Class"] = mainClassName
+        attributes["Class-Path"] = configurations["runtimeClasspath"].map { it.name }.joinToString(" ")
+        attributes["Implementation-Title"] = "Terasology-" + project.name
+        attributes["Implementation-Version"] = """${env["BUILD_NUMBER"]}, ${env["GIT_BRANCH"]}, ${env["BUILD_ID"]}"""
     }
 }
 
 configurations {
-    // for fetching modules for running a server
-    modules {
-        transitive = false
+    register("modules") {
+        description = "for fetching modules for running a server"
+        isTransitive = false
     }
 }
 
-run {
+tasks.named<JavaExec>("run") {
     description = "Run 'Terasology' to play the game as a standard PC application"
     group = "terasology run"
 
-    args "-homedir", "-noCrashReport"
-    jvmArgs "-Xmx1536m"
-    workingDir rootDir
+    args = listOf("-homedir", "-noCrashReport")
+    jvmArgs = listOf("-Xmx1536m")
+    workingDir = rootDir
 
-    dependsOn rootProject.moduleClasses
+    dependsOn(":moduleClasses")
```

Comment:
Hm, so everything else seems to become more (type-)safe, while you reference the dependency via a plain string there? Is there another way to do this in the Kotlin DSL?

Reply:
There's some tension going on in gradle-config-land. In one direction, you have Kotlin making things more type-safe. In the other direction, Gradle has this configuration-avoidance initiative to do lazy evaluation for certain types of things so it doesn't have to figure out the _entire_ multi-project configuration tree before it's able to run a single task. That's _discouraged_ direct links between sub-projects in many ways.

on second thought, I'm not sure how much the configuration-avoidance API has to do with this specific example.

from Kotlin's perspective, these are instances of Project. Project, as a Class, doesn't have a property named `moduleClasses`. build.gradle.kts files _configure_ projects but they do not define _new classes_ for each of them. Things like `rootProject.moduleClasses` only work with dynamic run-time attribute-lookup trickery.

We certainly could say something along the lines of `rootProject.tasks.get("moduleClasses")` but that's not any different than what we have here in terms of providing assurance that "moduleClasses" is a safe reference.

...and also the Gradle API is really big and provides a bunch of ways to get references to things and I certainly haven't learned everything about what's supposed to be the current best practice.

Reply:
Long-term, the “safe” way involves not knowing about the _tasks_ of other projects at all. You declare a dependency like "I depend on things from `"org.terasology.engine:engine"` and need assets of type `<Jar>` that conform to `bytecode_level=JAVA_1_8`" or something like that, the dependency-resolver does its thing and checks to see if that project exports anything that matches those criteria.